### PR TITLE
Feature/compactable instances

### PR DESCRIPTION
**Pull request by raveclassic**
Part of https://github.com/gcanti/fp-ts/pull/485
Adds instances of `Compactable` for `Array`, `Set`, `Option`, `StrMap` and `getCompactable` helper to `Either` and `Validation`

### CODE REVIEW

**Comment by gcanti** [comment]
@raveclassic could you please rebase on master rather than merging

**Comment by raveclassic** [comment]
@gcanti Fixed circular reference by moving `compactDefault` to `Option.ts` and `separateDefault` to `Either.ts`. That's the only way I think because these helpers require `Option` and `Either` classes. Since these helpers are likely to be autoimported most of the time by IDE/Editor I think it's ok to store them in `Option.ts` and `Either.ts` modules, what do you think?

**Comment by gcanti** [comment]
@raveclassic looks like the build was failing due to bad formatting of `test/Validation.ts`

```
...
> fp-ts@1.6.2 prettier /home/travis/build/gcanti/fp-ts
> prettier --no-semi --single-quote --print-width 120 --parser typescript --list-different "{src,test,examples,exercises}/**/*.ts"
test/Validation.ts
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! fp-ts@1.6.2 prettier: `prettier --no-semi --single-quote --print-width 120 --parser typescript --list-different "
...
```

(from https://travis-ci.org/gcanti/fp-ts/builds/394664387?utm_source=github_status&utm_medium=notification)

Anyway, isn't the circular dependencies just caused by importing the helper `separated` in `src/Either.ts` and `src/Option.ts`? If that's the case
you could just build a `Separated` by hand wihtout using the helper so that all the importa are just types

**Comment by raveclassic** [comment]
@gcanti You're right! 
I moved defaults back to Compactable and all tests even with circular reference. I think it's working because `separated` isn't accessed on module-level but in runtime so it's able to be resolved correctly.

**Comment by gcanti** [comment]
@raveclassic ok but there's still an issue with `separated`: every module that imports that small helper is now also importing the whole module `Option` and the whole module `Either` (via `Compactable`)

So far I carefully tried to import only types unless strictly necessary. Also I don't like having runtime circular dependencies, even if it seems to not cause issues (for now..)

Do we really need that helper? I don't see inference problems for the type checker (as opposed to tuple types, hence the `tuple` helper).
Can we at least move it into the `function` module?

**Comment by raveclassic** [comment]
@gcanti Fair enough. 

>Do we really need that helper? I don't see inference problems for the type checker (as opposed to tuple types, hence the tuple helper).

Yeah I think it can be removed because the only reason to add it was brevity.

File: src/Set.ts [hunk 1]
```diff
@@ -182,21 +204,24 @@ export const intersection = <A>(S: Setoid<A>): ((x: Set<A>, y: Set<A>) => Set<A>
  * @function
  * @since 1.2.0
  */
-export const partitionMap = <A, L, R>(x: Set<A>, f: (a: A) => Either<L, R>): { left: Set<L>; right: Set<R> } => {
+export const partitionMap = <A, L, R>(x: Set<A>, f: (a: A) => Either<L, R>): Separated<Set<L>, Set<R>> => {
```
Comment (gcanti) [inline src/Set.ts]: this is not related to this PR but IMO this signature is wrong, should be 

```ts
export const partitionMap = <L, R>(SL: Setoid<L>, SR: Setoid<R>) => <A>(x: Set<A>, f: (a: A) => Either<L, R>): Separated<Set<L>, Set<R>>
```

File: src/Set.ts [hunk 2]
```diff
@@ -292,3 +317,46 @@ export const fromArray = <A>(S: Setoid<A>) => (as: A[]): Set<A> => {
   }
   return r
 }
+
+const compact = <A>(fa: Set<Option<A>>): Set<A> => {
```
Comment (gcanti) [inline src/Set.ts]: Should be

```ts
const compact = <A>(S: Setoid<A>) => (fa: Set<Option<A>>): Set<A> =>
```

File: src/Set.ts [hunk 3]
```diff
@@ -292,3 +317,46 @@ export const fromArray = <A>(S: Setoid<A>) => (as: A[]): Set<A> => {
...
+  // tslint:disable:no-conditional-assignment
+  while (!(e = values.next()).done) {
+    const optionA = e.value
+    if (optionA.isSome()) {
+      result.add(optionA.value)
+    }
+  }
+  return result
+}
+const separate = <RL, RR>(fa: Set<Either<RL, RR>>): Separated<Set<RL>, Set<RR>> => {
```
Comment (gcanti) [inline src/Set.ts]: Should be

```ts
const separate = <A, B>(SA: Setoid<A>, SB: Setoid<B>) => (fa: Set<Either<A, B>>): Separated<Set<A>, Set<B>> =>
```

File: src/Set.ts [hunk 4]
```diff
@@ -292,3 +317,46 @@ export const fromArray = <A>(S: Setoid<A>) => (as: A[]): Set<A> => {
...
+    left,
+    right
+  }
+}
+
+/**
+ * @instance
+ * @since 1.7.0
+ */
+export const set: Compactable1<URI> = {
```
Comment (gcanti) [inline src/Set.ts]: this is not possible due to previous observations

File: src/Set.ts (hunk 1 above)
Comment (raveclassic) [reply src/Set.ts]: But why? I was also wondering why Set needs a Setoid to check the existence in current implementation. JavaScript Set implementation already filters out duplicates so you can't create `new Set([1, 1])`, result will always be `new Set([1])`. The same applies for `Set.prototype.add`

File: src/Option.ts [hunk 5]
```diff
@@ -555,6 +556,26 @@ export const fromRefinement = <A, B extends A>(refinement: Refinement<A, B>) =>
   return refinement(a) ? some(a) : none
 }
 
+const compact = <A>(fa: Option<Option<A>>): Option<A> => fa.chain(identity)
+const separate = <RL, RR, A>(fa: Option<Either<RL, RR>>): Separated<Option<RL>, Option<RR>> =>
```
[the remark this answers is not in the document]
Comment (raveclassic) [reply src/Option.ts]: fixed

File: src/Set.ts (hunk 1 above)
Comment (raveclassic) [reply src/Set.ts]: Furthermore `Filterable` instance requires `Set` to be a valid `Functor` without additional `Setoid`'s. And all tests pass for `map` without `Setoid`

Comment (gcanti) [reply src/Set.ts]: Because equality is a serious matter, actually as a mathematician I think that equality is the only thing that really matters :)

`Set` uses implicitly the following setoid

```ts
const setoidEvil: Setoid<any> = {
  equals: (x, y) => x === y
}
```

which is useless, try this `new Set([{ a: 1 }, { a: 1 }])`. 

Actually is worse than useless, is evil, because if you use that setoid, you are pretty much out of funcitonal programming (unless you add some machinery everytime you define a struct/function).

Let's define a `Person` struct

```ts
interface Person {
  name: string
}

const p1: Person = { name: 'foo' }
const p2: Person = { name: 'foo' }
```

"obviously" `p1` and `p2` are equal, however

```ts
console.log(setoidEvil.equals(p1, p2)) // => false
```

Even worse, consider the following function `person`

```ts
// constructor for Person
const person = (name: string): Person => ({ name })

const x1 = 'foo'
const x2 = 'foo'
const y1 = person(x1)
const y2 = person(x2)

console.log(setoidEvil.equals(x1, x2)) // => true
// but...
console.log(setoidEvil.equals(y1, y2)) // => false
```

as you can see `person` is not a pure function, which is (in practice) absurd.

So basically everytime you define an interface or a function or whatever, you are also defining an implicit setoid (or more setoids) instance.
Normally keeping those setoids implicit isn't much of a problem, but in the case of `Set` they are important and we should make them explicit, otherwise you can get duplicates (or even worse results).

Comment (raveclassic) [reply src/Set.ts]: Totally makes sense!
What about implementing it similarly to `Either`/`Validation` and accept a setoid in `getCompactable`?

Comment (raveclassic) [reply src/Set.ts]: @gcanti I've updated `Set`, take a look

Comment (raveclassic) [reply src/Set.ts]: @gcanti What do you think about such approach in general? When you can define all type arguments for a typeclass? I mean that `CompactableA`/`Compactable1A` which is similar to `Compactable2C`. What do you think about it? Is it ok? Is naming appropriate?

Comment (gcanti) [reply src/Set.ts]: I'm not sure, it doesn't feel right (and there's no equivalent in purescript-filterable). 
Likely there's no `Compactable` instance for `Set` just as there's no `Functor` instance.
For now I propose to either 

- remove `Set` from this PR and rediscuss later
- or remove `getCompactable` and define `compact` and `separate` as

```ts
export const compact = <A>(S: Setoid<A>) => (fa: Set<Option<A>>): Set<A> => ...
export const separate = <A, B>(SA: Setoid<A>, SB: Setoid<B>) => (fa: Set<Either<A, B>>): Separated<Set<A>, Set<B>> => ...
```

Comment (raveclassic) [reply src/Set.ts]: Ok, let's revert Set now and come back to it later.

Comment (raveclassic) [reply src/Set.ts]: @gcanti Ready

**Comment by gcanti** [comment]
@raveclassic I missed `CompactableA` and `Compactable1A` in my review (they are still there, I think we should remove them)

**Comment by raveclassic** [comment]
@gcanti Oops, are you ok to see them removed in the next PR with `Filterable`?

**Comment by gcanti** [comment]
np
